### PR TITLE
[wip] Aggregated Discovery: genericapiserver endpoint

### PR DESCRIPTION
**Pull request by alexzielenski**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

this PR adds an endpoint which advertises native resource types under the /discovery/v2 endpoint as documented in the KEP.

the endpoint is only available when the aggregated discovery feature gate is enabled

depends on:

#111978
#112112

/triage accepted
/sig api-machinery
/cc @Jefftree @apelisse @seans3 
/assign @Jefftree 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3352-aggregated-discovery
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/installer.go [hunk 1]
```diff
@@ -68,6 +68,72 @@ type action struct {
...
+				Kind:    r.Kind,
+			},
+			Verbs:      r.Verbs,
+			ShortNames: r.ShortNames,
+			Categories: r.Categories,
+		})
+		rootResources[r.Resource] = &apiResourceList[len(apiResourceList) - 1]
+	}
+
+	// Loop through all subrouces
```
Comment (Jefftree) [inline staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:111]: subresources

File: staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v2/handler.go [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) WebService() *restful.WebService {
+	mediaTypes, _ := negotiation.MediaTypesForSerializer(rdm.serializer)
+	ws := new(restful.WebService)
+	ws.Path(DiscoveryEndpointRoot)
+	ws.Doc("get available API groupversions and resources")
+
+	ws.Route(ws.GET("/v2").To(func(req *restful.Request, resp *restful.Response) {
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v2/handler.go:230]: v2 stands out here.  the API is metav1, right?

File: staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v2/handler.go [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+
+func (rdm *resourceDiscoveryManager) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	rdm.lock.RLock()
+	response := rdm.cachedResponse
+	etag := rdm.cachedResponseETag
+
+	// The cachedResponse is wiped out every time there might be a change to its
+	// contents.
+	if response != nil {
+		defer rdm.lock.RUnlock()
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v2/handler.go:249]: The need to hold a readlock while `ServeHTTPWithETag` isn't obvious as I read top to bottom here.  Are we protecting something in `ServeHTTPWithETag` or are we just protecting the response/etag reads above this block?

File: staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v2/handler.go [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+
+		// Document does not exist, recreate it
+		rdm.lock.Lock()
+		defer rdm.lock.Unlock()
+
+		// Now that we have taken exclusive lock, check to see if another thread
+		// recreated the document while we were waiting for the lock
+		response, etag = rdm.cachedResponse, rdm.cachedResponseETag
+		if response == nil {
+			// Re-order the apiGroups by their insertion order
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v2/handler.go:261]: Is a hold over from our previous means of prioritizing groups?  I thought we hardcoded a list somewhere to determine priority.

File: staging/src/k8s.io/apiserver/pkg/endpoints/installer.go [hunk 5]
```diff
@@ -68,6 +68,72 @@ type action struct {
...
+
+		parent, exists := rootResources[r.Resource]
+		if !exists {
+			// Had a subresource but not its parent???
+			continue
+		}
+
+		parent.Subresources = append(parent.Subresources, metav1.APISubresourceDiscovery{
+			Subresource: r.Subresource,
+			ReturnType: &metav1.APIDiscoveryKind{
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:126]: I need to chase this back  through.  Not all subresources have a return type. For instance, pods/exec.

Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:126]: I think the existing discovery doc has a quirk where if the subresource doesn't have a defined return gvk, the parent resource is assumed.

Comment (Jefftree) [reply staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:126]: ack, didn't realize there was the possibility that the current discovery document is incorrect. I'll look into this. We also need to fill in AcceptedTypes on certain subresources.

File: staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v2/handler.go (hunk 4 above)
Comment (alexzielenski) [reply staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v2/handler.go:261]: right. that list ends up not getting used anyway due to aggregator. ive removed it.

File: staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v2/handler.go (hunk 2 above)
Comment (alexzielenski) [reply staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v2/handler.go]: right. changed to v1

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 6]
```diff
@@ -676,11 +680,20 @@ func (s *GenericAPIServer) installAPIResources(apiPrefix string, apiGroupInfo *A
...
 
-		r, err := apiGroupVersion.InstallREST(s.Handler.GoRestfulContainer)
+		discoveryAPIResources, r, err := apiGroupVersion.InstallREST(s.Handler.GoRestfulContainer)
+
 		if err != nil {
 			return fmt.Errorf("unable to setup API %v: %v", apiGroupInfo, err)
 		}
 		resourceInfos = append(resourceInfos, r...)
+
+		s.DiscoveryResourceManager.AddGroupVersion(
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:690]: gated

**Review by deads2k — CHANGES_REQUESTED** [review]
largely good.  Once the earlier PRs merge (feature gate and API), I'll resweep for test coverage.

I expected the featuregate to be applied in more places and the subresources also need to be worked out.

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 7]
```diff
@@ -894,6 +901,11 @@ func installAPI(s *GenericAPIServer, c *Config) {
 	if c.EnableDiscovery {
 		s.Handler.GoRestfulContainer.Add(s.DiscoveryGroupManager.WebService())
 	}
+
+	if discoveryService := s.DiscoveryResourceManager.WebService(); discoveryService != nil {
```
[the remark this answers is not in the document]
Comment (Jefftree) [reply staging/src/k8s.io/apiserver/pkg/server/config.go:905]: On line 669, we create a noop resource manager if the feature gate is disabled.

**Pull request closed without merging**